File: libraries/android/sdk/src/main/java/com/voidhash/sdk/analytics/AnalyticsSessionManager.kt
package com.voidhash.sdk.analytics

import com.voidhash.sdk.cache.CacheManager
import org.json.JSONObject
import java.util.UUID

/** Cache key of the analytics session, shared with the other Voidhash SDKs. */
internal const val ANALYTICS_SESSION_CACHE_KEY = "voidhash:analytics:session"

/** 30 minutes — the inactivity window after which a new session starts. */
internal const val ANALYTICS_SESSION_TIMEOUT_MS = 30L * 60 * 1000

/**
 * Owns the analytics session id every captured event is stamped with.
 *
 * A session is a lowercase UUID that lives as long as events keep arriving
 * within [ANALYTICS_SESSION_TIMEOUT_MS] of each other. The id and the time of
 * the last event are persisted as `{id, lastEventAt}` under
 * [ANALYTICS_SESSION_CACHE_KEY] on every capture, so a session survives a
 * process restart as long as the app comes back within the window.
 *
 * All access is serialized, so concurrent captures never mint two sessions.
 */
class AnalyticsSessionManager(
    private val cacheManager: CacheManager,
    private val clock: () -> Long = { System.currentTimeMillis() },
    private val uuidFactory: () -> String = { UUID.randomUUID().toString() },
    private val onWarning: (String) -> Unit = {},
) {
    private data class Session(val id: String, val lastEventAt: Long)

    private var session: Session? = null
    private var loaded = false

    /**
     * Returns the session id for an event captured now: rotates when there is
     * no session or the last event is older than the timeout, then records the
     * event time and persists the session.
     */
    @Synchronized
    fun current(): String {
        val now = clock()
        val live = liveSession(now) ?: Session(uuidFactory().lowercase(), now)
        persist(live.copy(lastEventAt = now))
        return live.id
    }

    /**
     * Returns the session id without extending the session. A missing or
     * expired session is replaced so the caller always sees the id the next
     * capture would carry; the replacement is not persisted until then.
     */
    @Synchronized
    fun peek(): String {
        val now = clock()
        liveSession(now)?.let { return it.id }
        val fresh = Session(uuidFactory().lowercase(), now)
        session = fresh
        return fresh.id
    }

    /** Starts a new session immediately, regardless of the timeout. */
    @Synchronized
    fun rotate(): String {
        val fresh = Session(uuidFactory().lowercase(), clock())
        persist(fresh)
        return fresh.id
    }

    private fun liveSession(now: Long): Session? {
        val existing = loadOnce() ?: return null
        if (now - existing.lastEventAt > ANALYTICS_SESSION_TIMEOUT_MS) return null
        return existing
    }

    private fun loadOnce(): Session? {
        if (loaded) return session
        loaded = true
        session = try {
            cacheManager.getObject(ANALYTICS_SESSION_CACHE_KEY)?.value?.let { stored ->
                val id = stored.optString("id")
                if (id.isEmpty() || !stored.has("lastEventAt")) null
                else Session(id, stored.getLong("lastEventAt"))
            }
        } catch (error: Throwable) {
            onWarning("Failed to read the analytics session: ${error.message}")
            null
        }
        return session
    }

    /** Keeps the in-memory session even when the cache write fails, so ids stay stable. */
    private fun persist(next: Session) {
        session = next
        loaded = true
        try {
            cacheManager.set(
                ANALYTICS_SESSION_CACHE_KEY,
                JSONObject().put("id", next.id).put("lastEventAt", next.lastEventAt),
            )
        } catch (error: Throwable) {
            onWarning("Failed to store the analytics session: ${error.message}")
        }
    }
}

File: libraries/ios/Sources/Voidhash/Analytics/AnalyticsSessionManager.swift
import Foundation
import VoidhashCore

/// Persisted analytics session, shared byte-for-byte with the React Native and Android SDKs.
struct AnalyticsSessionRecord: Codable, Sendable, Equatable {
    /// Lowercase UUID string.
    let id: String
    /// Millisecond epoch timestamp of the last event captured in this session.
    let lastEventAt: Double
}

/// Owns the analytics session id stamped on every captured event.
///
/// A session ends after 30 minutes without a capture; the next capture starts a new one. The
/// record is persisted on every capture, so a session survives a process restart as long as
/// the app comes back within the timeout.
actor AnalyticsSessionManager {
    /// Cache key of the persisted ``AnalyticsSessionRecord``.
    static let cacheKey = "voidhash:analytics:session"
    /// Inactivity after which the next capture starts a new session.
    static let inactivityTimeoutMilliseconds: Double = 30 * 60 * 1000

    private let cacheManager: CacheManager
    private let now: @Sendable () -> Double
    private let makeSessionId: @Sendable () -> String
    private var record: AnalyticsSessionRecord?
    private var hasLoaded = false

    /// - Parameters:
    ///   - cacheManager: Storage of the session record.
    ///   - now: Millisecond epoch clock, injectable for tests.
    ///   - makeSessionId: Session id factory, injectable for tests.
    init(
        cacheManager: CacheManager,
        now: @escaping @Sendable () -> Double = { Date().timeIntervalSince1970 * 1000 },
        makeSessionId: @escaping @Sendable () -> String = { UUID().uuidString.lowercased() }
    ) {
        self.cacheManager = cacheManager
        self.now = now
        self.makeSessionId = makeSessionId
    }

    /// Returns the session id for an event captured now, starting a new session when the
    /// previous one timed out, and records the capture time.
    func current() async -> String {
        let timestamp = now()
        let active = await activeRecord(at: timestamp)
        let next = AnalyticsSessionRecord(id: active?.id ?? makeSessionId(), lastEventAt: timestamp)
        await persist(next)
        return next.id
    }

    /// Returns the id the next capture will carry without counting as activity. A timed-out
    /// session is replaced, so the returned id is never stale.
    func peek() async -> String {
        let timestamp = now()
        if let active = await activeRecord(at: timestamp) {
            return active.id
        }
        return await rotate()
    }

    /// Starts a new session unconditionally and returns its id.
    @discardableResult
    func rotate() async -> String {
        let next = AnalyticsSessionRecord(id: makeSessionId(), lastEventAt: now())
        await persist(next)
        return next.id
    }

    private func activeRecord(at timestamp: Double) async -> AnalyticsSessionRecord? {
        if !hasLoaded {
            record = await cacheManager.get(
                AnalyticsSessionManager.cacheKey, as: AnalyticsSessionRecord.self)?.value
            hasLoaded = true
        }
        guard let record,
            timestamp - record.lastEventAt <= AnalyticsSessionManager.inactivityTimeoutMilliseconds
        else {
            return nil
        }
        return record
    }

    private func persist(_ next: AnalyticsSessionRecord) async {
        record = next
        hasLoaded = true
        await cacheManager.set(AnalyticsSessionManager.cacheKey, value: next)
    }
}
